[ADD] Added new functionality to generate pos voucher from pos screen.

janiniravc opened the pull request:
Task URL: https://www.druidoo.io/web#id=1405&action=367&model=project.task&view_type=form&menu_id=237

devel-zs committed:
[ADD] Added new functionality to generate pos voucher from pos screen.

ivantodorovich commented on pos_voucher/models/pos_voucher.py:
```diff
@@ -11,6 +11,9 @@ class PosVoucherType(models.Model):
     sequence_id = fields.Many2one('ir.sequence', required=True)
+    product_id = fields.Many2one('product.product', 'Product',
+                                 domain="[('available_in_pos', '=', True),\
+                                 ('type', '=', 'service')]")
```
This is a lot more readable:

```py
    product_id = fields.Many2one(
        'product.product',
        'Product',
        domain=[('available_in_pos', '=', True), ('type', '=', 'service')],
    )
```

Please look at https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#models for code style conventions

ivantodorovich commented on pos_voucher/models/pos_voucher.py:
```diff
@@ -123,6 +126,33 @@ def get_voucher_by_code(self, code, partner_id):
+        ret_list = []
+        if pos_ref:
+            pos_order = self.env['pos.order'].search([('pos_reference',
+                                                       '=', pos_ref)])
```
Same here:

```py
pos_order = self.env['pos.order'].search([
    ('pos_reference', '=', pos_ref)])
```

ivantodorovich commented on pos_voucher/static/src/js/pos.js:
```diff
@@ -234,4 +236,83 @@ models.load_fields("account.bank.statement", ['is_voucher']);
+                        self.lock_screen(false);
+                    }, 1000);
+                    self.print_web();
+                    //window.print();
```
```suggestion
```

ivantodorovich commented on pos_voucher/static/src/js/pos.js:
```diff
@@ -234,4 +236,83 @@ models.load_fields("account.bank.statement", ['is_voucher']);
+                    setTimeout(function(){
+                        self.lock_screen(false);
+                    }, 1000);
+                    self.print_web();
```
Will it work if the POS is using PosBox to print it with a thermal printer? :thinking: 

Why don't you add the vouchers in the same report than the receipt, so that they're printed at the very same time and you don't have to worry about making it compatible with posbox?

ivantodorovich replied:
This is very important.

I've just tested on a database with these settings, and the print window is not even shown, so it's impossible to print the vouchers.

I think we really need to integrate the vouchers into the Receit report

![image](https://user-images.githubusercontent.com/1914185/73834108-68e68780-480b-11ea-9ae1-d5e4d7d28537.png)

ivantodorovich commented on pos_voucher/static/src/js/screens.js:
```diff
@@ -191,6 +192,52 @@ var voucherListScreenWidget = screens.ScreenWidget.extend({
 gui.define_screen({name:'voucherlist', widget: voucherListScreenWidget});
 
+
+var popup_voucher_generate = PopupWidget.extend({
```
```suggestion
var popupVoucherGenerate = PopupWidget.extend({
```

ivantodorovich commented on pos_voucher/static/src/js/screens.js:
```diff
@@ -191,6 +192,52 @@ var voucherListScreenWidget = screens.ScreenWidget.extend({
+
+
 return {
     voucherListScreenWidget: voucherListScreenWidget,
```
```suggestion
    voucherListScreenWidget: voucherListScreenWidget,
    popupVoucherGenerate: popupVoucherGenerate,
```

We want to return this object so that future modules can extend it

ivantodorovich commented on pos_voucher/static/src/xml/pos.xml:
```diff
@@ -95,4 +95,77 @@
+        </div>
+    </t>
+
+    <t t-name="popup_voucher_generate">
```
popupVoucherGenerate

devel-zs committed:
[IMP]Added code for Automatic Receipt Printing for pos voucher

devel-zs committed:
[IMP] pycodestyle resolved.

ivantodorovich commented:
@janiniravc can you fix that the vouchers generated are not validated?
![image](https://user-images.githubusercontent.com/1914185/74170382-1a7f2180-4c2d-11ea-914b-456cff8edba2.png)

They should be validated with the order, as soon as they're printed

devel-zs committed:
[MDY] call validate method when create pos voucher from pos screen.

ivantodorovich committed:
[MOV] templates.xml -> assets.xml

ivantodorovich committed:
[IMP] Code and usability

ivantodorovich committed:
[IMP] Print vouchers right after the ticket, and code cleanup

ivantodorovich committed:
[IMP] Block cancellation of vouchers created from POS orders

ivantodorovich committed:
[IMP] code: action_draft

[FIX] Import ValidationError

[IMP] Split .py files

[IMP] Use ValidationError instead of UserError

[FIX] LINT

ivantodorovich committed:
[FIX] _create_pos_vouchers()

ivantodorovich committed:
[I18N] Translations

ivantodorovich committed:
[I18N] Translations
